optuna/integration/pytorch_lightning.py

```diff
@@ -56,3 +59,84 @@ def on_validation_end(self, trainer: Trainer, pl_module: LightningModule) -> Non
         if self._trial.should_prune():
             message = "Trial was pruned at epoch {}.".format(epoch)
             raise optuna.TrialPruned(message)
+
+
+class PyTorchLightningDDPPruningCallback(Callback):
+    """PyTorch Lightning DDP callback to prune unpromising trials.
+
+    See `the example <https://github.com/optuna/optuna-examples/blob/
+    main/pytorch/pytorch_lightning_simple.py>`__
+    if you want to add a pruning callback which observes accuracy.
+
+    Args:
+        trial:
+            A :class:`~optuna.trial.Trial` corresponding to the current evaluation of the
+            objective function.
+        monitor:
+            An evaluation metric for pruning, e.g., ``val_loss`` or
+            ``val_acc``. The metrics are obtained from the returned dictionaries from e.g.
+            ``pytorch_lightning.LightningModule.training_step`` or
+            ``pytorch_lightning.LightningModule.validation_epoch_end`` and the names thus depend on
+            how this dictionary is formatted.
+    """
+
+    def __init__(self, trial: optuna.trial.Trial, monitor: str) -> None:
+        _imports.check()
+
+        if not isinstance(trial.study._storage, _CachedStorage):
+            raise ValueError("PyTorchLightningDDPPruningCallback supports only RDB.")
+
+        super().__init__()
+
+        self._trial = trial
+        self.monitor = monitor
+
+    def on_validation_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
+        epoch = pl_module.current_epoch
+
+        current_score = trainer.callback_metrics.get(self.monitor)
+        if current_score is None:
+            message = (
+                "The metric '{}' is not in the evaluation logs for pruning. "
+                "Please make sure you set the correct metric name.".format(self.monitor)
+            )
+            warnings.warn(message)
+            return
+
+        # When multi gpu, on_validation_end function executes multi times.
+        # So Exception is ignored to avoid IntegrityError.
+        distributed_backend = trainer.accelerator_connector.distributed_backend
+        try:
+            self._trial.report(current_score, step=epoch)
+        except sqlalchemy.exc.IntegrityError:
+            if distributed_backend is not None:
+                pass
+            else:
+                raise
+
+        if self._trial.should_prune():
+            trainer.should_stop = True
+            try:
+                self._trial.set_user_attr("pruned", True)
+                self._trial.set_user_attr("epoch", epoch)
+            except sqlalchemy.exc.IntegrityError:
+                if distributed_backend is not None:
+                    pass
+                else:
+                    raise
+
+    # Because on_validation_end is executed in threads, when on_fit_end is executed,
+    # it is necessary to report an objective function value for a given step via RDB.
```

Comment:
[Question]

I want to know about re-reporting. In my understanding, some `self._trial.report` calls succeed and some fail in `on_validation._end`. Is there possibility of reporting the same value multiple times? (Sorry my poor understanding)
https://github.com/optuna/optuna/pull/2849/files#diff-8fc58e871098cfca85d84c185999114d609aecd1f5dd6ed8cb25e172bc996272R110

Reply:
_trial_report are executed on spawned processes, so memory's trial(intermediate_value in each step) in main process is not updated. _trial_report is executed to update the memory, not to update the DB.

Reply:
I see. Thank you for clarifying!